src/Features/CSharp/Portable/BraceCompletion/CurlyBraceCompletionService.cs

```diff
@@ -0,0 +1,461 @@
+                newLineEdit = new TextChange(new TextSpan(closingPoint - 1, 0), newLineString);
+                textToFormat = originalDocumentText.WithChanges(newLineEdit.Value);
+
+                // Modify the closing point location to adjust for the newly inserted line.
+                closingPoint += newLineString.Length;
+            }
+
+            // Format the text that contains the newly inserted line.
+            var (formattingChanges, newClosingPoint) = await FormatTrackingSpanAsync(
+                document.WithText(textToFormat),
+                documentOptions,
+                openingPoint,
+                closingPoint,
+                shouldHonorAutoFormattingOnCloseBraceOption: false,
+                braceFormattingIndentationRules: GetBraceIndentationFormattingRules(documentOptions),
+                cancellationToken).ConfigureAwait(false);
+            closingPoint = newClosingPoint;
+            var formattedText = textToFormat.WithChanges(formattingChanges);
+
+            // Get the empty line between the curly braces.
+            var desiredCaretLine = GetLineBetweenCurlys(closingPoint, formattedText);
+            Debug.Assert(desiredCaretLine.GetFirstNonWhitespacePosition() == null, "the line between the formatted braces is not empty");
+
+            // Set the caret position to the properly indented column in the desired line.
+            var newDocument = document.WithText(formattedText);
+            var newDocumentText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var caretPosition = GetIndentedLinePosition(newDocument, newDocumentText, desiredCaretLine.LineNumber, cancellationToken);
+
+            // The new line edit is calculated against the original text, d0, to get text d1.
+            // The formatting edits are calculated against d1 to get text d2.
+            // Merge the formatting and new line edits into a set of whitespace only text edits that all apply to d0.
+            var overallChanges = newLineEdit != null ? GetMergedChanges(newLineEdit.Value, formattingChanges, formattedText) : formattingChanges;
+            return new BraceCompletionResult(overallChanges, caretPosition);
+
+            static bool HasExistingEmptyLineBetweenBraces(int openingPoint, int closingPoint, SourceText sourceText)
+            {
+                // Check if there is an empty new line between the braces already.  If not insert one.
+                // This handles razor cases where they insert an additional empty line before calling brace completion.
+                var openingPointLine = sourceText.Lines.GetLineFromPosition(openingPoint).LineNumber;
+                var closingPointLine = sourceText.Lines.GetLineFromPosition(closingPoint).LineNumber;
+
+                return closingPointLine - 1 > openingPointLine && sourceText.Lines[closingPointLine - 1].IsEmptyOrWhitespace();
```

Comment:
should we be checking that there is *exactly* one empty line between the braces?  currently, this allows for many lines, where only the one before the close brace is empty.

Reply:
could use GetLineBetweenCurlys here